apollo-mockserver/src/jvmMain/kotlin/com/apollographql/apollo3/mockserver/MockServer.kt

```diff
@@ -1,65 +1,80 @@
-  private val mockWebServer = MockWebServer().apply { dispatcher = mockServerHandler.toOkHttpDispatcher() }
+@Suppress("BlockingMethodInNonBlockingContext")
+actual class MockServer actual constructor(
+    override val mockServerHandler: MockServerHandler,
+) : MockServerInterface {
+  private val serverSocket = ServerSocket(0)
+  private val mockRequests = mutableListOf<MockRequest>()
 
-  override fun enqueue(mockResponse: MockResponse) {
-    (mockServerHandler as? QueueMockServerHandler)?.enqueue(mockResponse)
-        ?: error("Apollo: cannot call MockServer.enqueue() with a custom handler")
-  }
+  private val dispatcher = Executors.newFixedThreadPool(5).asCoroutineDispatcher()
+  private val coroutineScope = CoroutineScope(SupervisorJob() + dispatcher)
 
-  override fun takeRequest(): MockRequest {
-    return mockWebServer.takeRequest(10, TimeUnit.MILLISECONDS)?.toApolloMockRequest() ?: error("No recorded request")
+  init {
+    coroutineScope.launch {
+      while (true) {
+        val clientSocket = try {
+          serverSocket.accept()
+        } catch (_: Exception) {
+          // An exception here means the server socket has been closed (stop() was called)
+          break
+        }
+        coroutineScope.launch { handleClient(clientSocket) }
+      }
+    }
   }
 
-  private fun RecordedRequest.toApolloMockRequest() = MockRequest(
-      method = method!!,
-      path = path!!,
-      version = parseRequestLine(requestLine).third,
-      headers = headers.toMap(),
-      body = body.peek().readByteString()
-  )
+  private suspend fun handleClient(clientSocket: Socket) {
+    try {
+      val clientSource = clientSocket.getInputStream().source().buffer()
+      val clientSink = clientSocket.getOutputStream().sink().buffer()
 
-  private fun MockResponse.toOkHttpMockResponse() = okhttp3.mockwebserver.MockResponse()
-      .setResponseCode(statusCode)
-      .apply {
-        this@toOkHttpMockResponse.headers.forEach {
-          addHeader(it.key, it.value)
-        }
-      }.setBody(Buffer().apply { write(body) })
-      .setHeadersDelay(delayMillis, TimeUnit.MILLISECONDS)
+      val mockRequest = readRequest(clientSource)!!
+      mockRequests += mockRequest
 
-  private fun MockServerHandler.toOkHttpDispatcher() = object : Dispatcher() {
-    override fun dispatch(request: RecordedRequest): okhttp3.mockwebserver.MockResponse {
-      val apolloMockRequest = request.toApolloMockRequest()
-      val mockResponse = try {
-        handle(apolloMockRequest)
-      } catch (e: Exception) {
-        throw Exception("MockServerHandler.handle() threw an exception: ${e.message}", e)
-      }
-      return mockResponse.toOkHttpMockResponse()
+      val mockResponse = mockServerHandler.handle(mockRequest)
+      delay(mockResponse.delayMillis)
+      writeResponse(clientSink, mockResponse, mockRequest.version)
+    } catch (_: CancellationException) {
+      // Ignored: this is expected when the MockServer is closed
```

Comment:
Gosh, I'll never get used to the `CancellationException` ... I thought this was the one that needed to be rethrown all the time for structured concurrency to work well?